File: PYTHON_GENERATE_PROFILES/RADIAL_PROFILE_GENERATION_V1001.py
# Brendan M Unikewicz, PhD Student
# MIT, Dept. Mechanical Engineering
# Date of Creation: 06162023
# Code Purpose: Generating profiles for microPump, i.e. controlPump.py

import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# User Inputs
aStart_cycling = float(input('Enter starting value for a_start (mm): '))
aStop_cycling = float(input('Enter stopping value for a_stop (mm): '))
nInterpolating = int(input('Enter the number of interpolating points (N): '))

ratesInfusion = [0.01, 0.02, 0.04, 0.08, 0.16, 0.32]  # mm/s

samplingRate = 10000  # Hz, this is non-important... just DSP-chosen value
dt = 1 / samplingRate

mm3_nL = 1000  # mm^3 to nL

fig, ax = plt.subplots(2, 1)
radialInfusion_profiles = []
timeVectors = []
volumeInfusion_profiles = []
timeInfuse_interpolateds = []
volumeInfuse_interpolateds = []
volInfuse_rates = []
volInfuse_mags = []
interpInfuses = []
actualInfuses = []
pDiff_infuses = []
infusionTimes = []

# Infusion Rates Generation
for rate in ratesInfusion:
    radialInfusion_profile = np.arange(aStart_cycling, aStop_cycling, dt*rate)
    timeVector = np.arange(1, len(radialInfusion_profile)+1) * dt
    volumeInfusion_profile = (4/3) * np.pi * radialInfusion_profile**3 * mm3_nL

    timeInfuse_interpolated = np.linspace(timeVector[0], timeVector[-1], nInterpolating)
    volumeInfuse_interpolated = np.interp(timeInfuse_interpolated, timeVector, volumeInfusion_profile)

    ax[1].plot(timeInfuse_interpolated, volumeInfuse_interpolated, 'ok', label=f'Rate: {rate} mm/s')

    volInfuse_rate = np.diff(volumeInfuse_interpolated) / np.diff(timeInfuse_interpolated)
    volInfuse_mag = np.diff(volumeInfuse_interpolated)

    interpInfuse = np.trapz(volumeInfuse_interpolated, timeInfuse_interpolated)
    actualInfuse = np.trapz(volumeInfusion_profile, timeVector)
    pDiff_infuse = ((interpInfuse - actualInfuse) / interpInfuse) * 100

    ax[0].plot(timeVector, radialInfusion_profile, label=f'Rate: {rate} mm/s')
    ax[1].plot(timeVector, volumeInfusion_profile, label=f'Rate: {rate} mm/s')

    infusionTime = timeVector[-1]

    radialInfusion_profiles.append(radialInfusion_profile)
    timeVectors.append(timeVector)
    volumeInfusion_profiles.append(volumeInfusion_profile)
    timeInfuse_interpolateds.append(timeInfuse_interpolated)
    volumeInfuse_interpolateds.append(volumeInfuse_interpolated)
    volInfuse_rates.append(volInfuse_rate)
    volInfuse_mags.append(volInfuse_mag)
    interpInfuses.append(interpInfuse)
    actualInfuses.append(actualInfuse)
    pDiff_infuses.append(pDiff_infuse)
    infusionTimes.append(infusionTime)

ax[0].grid(True, which='both', linestyle='--', linewidth=0.5)
ax[1].grid(True, which='both', linestyle='--', linewidth=0.5)
ax[0].set_title('Infusion Radial (top) & Volumetric (bottom) Rates')
plt.show()

# Writing Infusion data to CSV
pd.DataFrame(volInfuse_mags).to_csv('infusionMags.csv', index=False, header=False)
pd.DataFrame(volInfuse_rates).to_csv('infusionRates.csv', index=False, header=False)
pd.DataFrame(infusionTimes).to_csv('infusionTimes.csv', index=False, header=False)

File: MATLAB_GENERATE_PROFILES/RADIAL_PROFILE_GENERATION_V1001.m
% Brendan M Unikewicz, PhD Student
% MIT, Dept. Mechanical Engineering
% Date of Creation: 06162023
% Code Purpose: Generating profiles for microPump, i.e. controlPump.py

clear; close all; format bank; 

% User Inputs
aStart_cycling = input('Enter starting value for a_start (mm): ');
aStop_cycling = input('Enter stopping value for a_stop (mm): ');
nInterpolating = input('Enter the number of interpolating points (N): ');

ratesInfusion = [0.01 0.02 0.04 0.08 0.16 0.32]; % mm/s

%10uLSyringe_maxRate = 657; % nL/s
%25uLSyringe_maxRate = 1482; % nL/s
%50uLSyringe_maxRate = 2963; % nL/s

samplingRate = 500; % Hz, this is non-important... just DSP-chosen value
dt = 1/samplingRate; 

mm3_nL = 1000; % mm^3 to nL

figure; 
% Infusion Rates Generation
for i = 1:length(ratesInfusion)
    radialInfusion_profile{i,:} = aStart_cycling:(dt*ratesInfusion(i)):aStop_cycling; 
    timeVector{i,:} = (1:1:length(radialInfusion_profile{i,:}))*dt; 
    volumeInfusion_profile{i,:} = (4/3) * pi * (radialInfusion_profile{i,1}).^3 * mm3_nL; 
                         
    timeInfuse_interpolated{i,:} = linspace(min(timeVector{i,:}),max(timeVector{i,:}),nInterpolating);
    volumeInfuse_interpolated{i,:} = interp1(timeVector{i,:},volumeInfusion_profile{i,:},timeInfuse_interpolated{i,:}) ;
    subplot(2,1,2)
    plot(timeInfuse_interpolated{i,:},volumeInfuse_interpolated{i,:},'ok'); hold on; 

    volInfuse_rate{i,:} = diff(volumeInfuse_interpolated{i,:}) ./ diff(timeInfuse_interpolated{i,:}); 
    volInfuse_mag{i,:} = diff(volumeInfuse_interpolated{i,:}); 

    % Integration check
    interpInfuse(i) = trapz(timeInfuse_interpolated{i,:},volumeInfuse_interpolated{i,:});
    actualInfuse(i) = trapz(timeVector{i,:},volumeInfusion_profile{i,:});
    pDiff_infuse(i) = ((interpInfuse(i) - actualInfuse(i)) / interpInfuse(i)) * 100; 

    subplot(2,1,1); title('Infusion Radial (top) & Volumetric (bottom) Rates'); 
    plot(timeVector{i,:},radialInfusion_profile{i,:},'LineWidth',1.5); hold on; 
    ylabel('Effective Cavity Radii (mm)'); xlabel('Time (s)')
    grid on; grid minor; % legend(['Rate: ',num2str(ratesInfusion(i)),'mm/s']); hold on; 
    subplot(2,1,2); 
    plot(timeVector{i,:},volumeInfusion_profile{i,:},'LineWidth',1.5); hold on; 
    ylabel('Infused Fluid (nL)'); xlabel('Time (s)')
    grid on; grid minor; % legend(['Rate: ',num2str(ratesInfusion(i)),'mm/s']); hold on;

    infusionTime(i) = max(timeVector{i,:}); 
end
hold off; 

% Writing Infusion data to CSV
filenameInfuse_rate = 'infusionRates.csv'; 
filenameInfuse_mag = 'infusionMags.csv'; 
csvwrite(filenameInfuse_mag, cell2mat(volInfuse_mag));
csvwrite(filenameInfuse_rate, cell2mat(volInfuse_rate));

% Writing Infusion time data to CSV
filenameInfuse_time = 'infusionTimes.csv'; 
csvwrite(filenameInfuse_time, infusionTime);
